instrumentation-genai/opentelemetry-instrumentation-google-genai/src/opentelemetry/instrumentation/google_genai/generate_content.py

```diff
@@ -646,25 +633,33 @@ async def instrumented_generate_content_stream(
         helper = _GenerateContentInstrumentationHelper(
             self, otel_wrapper, model
         )
-        with helper.start_span_as_current_span(
-            model, "google.genai.AsyncModels.generate_content_stream"
-        ):
-            helper.process_request(contents, config)
-            try:
-                async for response in await wrapped_func(
-                    self,
-                    model=model,
-                    contents=contents,
-                    config=config,
-                    **kwargs,
-                ):  # pyright: ignore
-                    helper.process_response(response)
-                    yield response  # pyright: ignore
-            except Exception as error:
-                helper.process_error(error)
-                raise
-            finally:
-                helper.finalize_processing()
+
+        async def _internal_generator():
+            with helper.start_span_as_current_span(
+                model, "google.genai.AsyncModels.generate_content_stream"
+            ):
+                helper.process_request(contents, config)
+                try:
+                    async for response in await wrapped_func(
+                        self,
+                        model=model,
+                        contents=contents,
+                        config=config,
+                        **kwargs,
+                    ):  # pyright: ignore
+                        helper.process_response(response)
+                        yield response  # pyright: ignore
+                except Exception as error:
+                    helper.process_error(error)
+                    raise
+                finally:
+                    helper.finalize_processing()
+
+        class _GeneratorProvider:
+            def __aiter__(self):
+                return _internal_generator()
+
+        return _GeneratorProvider()
```

Comment:
Looking at https://googleapis.github.io/python-genai/genai.html#genai.models.AsyncModels.generate_content_stream

I think this code changes the semantics of the wrapped function slightly. Now when the user calls `await client.aio.models.generate_content_stream()`, the `wrapped_func` still has not been called and the span has not started.

Something like this should work but haven't tested it:

```py
async def instrumented_generate_content_stream(...):
  with helper.start_span_as_current_span(...):
    gen = await wrapped_func(...)
 
    async def _internal_generator():
      async for response in gen:
        ...
  
    return _internal_generator()
```

Reply:
Good point. It was a bit more complicated than that, but updated based on this feedback to begin eager execution earlier, before the point at which the responses are being consumed.

Reply:
Marking resolved given the restructuring here to eagerly begin the processing.